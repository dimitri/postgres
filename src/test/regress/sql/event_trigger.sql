-- should fail, return type mismatch
create event trigger regress_event_trigger
   on ddl_command_start
   execute procedure pg_backend_pid();

-- OK
create function test_event_trigger() returns event_trigger as $$
BEGIN
    RAISE NOTICE 'test_event_trigger: % % %', tg_event, tg_context, tg_tag;
    RAISE NOTICE 'test_event_trigger: %, %', tg_operation, tg_obtypename;
    RAISE NOTICE 'test_event_trigger: %', tg_command;
END
$$ language plpgsql;

-- should fail, no elephant_bootstrap entry point
create event trigger regress_event_trigger on elephant_bootstrap
   execute procedure test_event_trigger();

-- OK
create event trigger regress_event_trigger_start on ddl_command_start
   execute procedure test_event_trigger();

create event trigger regress_event_trigger_trace on ddl_command_trace
   when tag in ('create table')
   execute procedure test_event_trigger();

-- test creating an event trigger on ddl_command_end
create event trigger regress_event_trigger_end on ddl_command_end
   when context in ('toplevel', 'generated')
   execute procedure test_event_trigger();

-- but we won't actually use it
drop event trigger regress_event_trigger_end;

-- should fail, food is not a valid filter variable
create event trigger regress_event_trigger2 on ddl_command_start
   when food in ('sandwhich')
   execute procedure test_event_trigger();

-- should fail, sandwhich is not a valid command tag
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('sandwhich')
   execute procedure test_event_trigger();

-- should fail, sandwhich is not a valid command context either
create event trigger regress_event_trigger2 on ddl_command_start
   when context in ('sandwhich')
   execute procedure test_event_trigger();

-- should fail, create skunkcabbage is not a valid comand tag
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('create table', 'create skunkcabbage')
   execute procedure test_event_trigger();

-- should fail, skunkcabbage is not a valid comand context
create event trigger regress_event_trigger2 on ddl_command_start
   when context in ('toplevel', 'skunkcabbage')
   execute procedure test_event_trigger();

-- should fail, can't have event triggers on event triggers
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('DROP EVENT TRIGGER')
   execute procedure test_event_trigger();

-- should fail, can't have same filter variable twice
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('create table') and tag in ('CREATE FUNCTION')
   execute procedure test_event_trigger();

-- OK
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('create table', 'CREATE FUNCTION')
    and context in ('toplevel', 'generated', 'query', 'subcommand')
   execute procedure test_event_trigger();

-- OK
comment on event trigger regress_event_trigger_start is 'test comment';

-- should fail, event triggers are not schema objects
comment on event trigger wrong.regress_event_trigger_start is 'test comment';

-- drop as non-superuser should fail
create role regression_bob;
set role regression_bob;
create event trigger regress_event_trigger_noperms on ddl_command_start
   execute procedure test_event_trigger();
reset role;

-- all OK
alter event trigger regress_event_trigger_start enable replica;
alter event trigger regress_event_trigger_start enable always;
alter event trigger regress_event_trigger_start enable;
alter event trigger regress_event_trigger_start disable;

-- regress_event_trigger2 should fire, but not regress_event_trigger
create table event_trigger_fire1 (a int);

-- but nothing should fire here
drop table event_trigger_fire1;

-- alter owner to non-superuser should fail
alter event trigger regress_event_trigger_start owner to regression_bob;

-- alter owner to superuser should work
alter role regression_bob superuser;
alter event trigger regress_event_trigger_start owner to regression_bob;

-- should fail, name collision
alter event trigger regress_event_trigger_start rename to regress_event_trigger2;

-- OK
alter event trigger regress_event_trigger_start rename to regress_event_trigger3;

-- should fail, doesn't exist any more
drop event trigger regress_event_trigger;

-- should fail, regression_bob owns regress_event_trigger2/3
drop role regression_bob;

-- these are all OK; the second one should emit a NOTICE
drop event trigger if exists regress_event_trigger2;
drop event trigger if exists regress_event_trigger2;
drop event trigger regress_event_trigger3;
drop event trigger regress_event_trigger_trace;
drop function test_event_trigger();
drop role regression_bob;
