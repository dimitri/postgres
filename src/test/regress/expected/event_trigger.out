-- should fail, return type mismatch
create event trigger regress_event_trigger
   on ddl_command_start
   execute procedure pg_backend_pid();
ERROR:  function "pg_backend_pid" must return type "event_trigger"
-- OK
create function test_event_trigger() returns event_trigger as $$
BEGIN
    RAISE NOTICE 'test_event_trigger: % %', tg_event, tg_tag;
END
$$ language plpgsql;
-- should fail, no elephant_bootstrap entry point
create event trigger regress_event_trigger on elephant_bootstrap
   execute procedure test_event_trigger();
ERROR:  unrecognized event name "elephant_bootstrap"
-- OK
create event trigger regress_event_trigger on ddl_command_start
   execute procedure test_event_trigger();
-- OK
create event trigger regress_event_trigger_end on ddl_command_end
   execute procedure test_event_trigger();
-- should fail, food is not a valid filter variable
create event trigger regress_event_trigger2 on ddl_command_start
   when food in ('sandwhich')
   execute procedure test_event_trigger();
ERROR:  unrecognized filter variable "food"
-- should fail, sandwhich is not a valid command tag
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('sandwhich')
   execute procedure test_event_trigger();
ERROR:  filter value "sandwhich" not recognized for filter variable "tag"
-- should fail, create skunkcabbage is not a valid comand tag
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('create table', 'create skunkcabbage')
   execute procedure test_event_trigger();
ERROR:  filter value "create skunkcabbage" not recognized for filter variable "tag"
-- should fail, can't have event triggers on event triggers
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('DROP EVENT TRIGGER')
   execute procedure test_event_trigger();
ERROR:  event triggers are not supported for "DROP EVENT TRIGGER"
-- should fail, can't have same filter variable twice
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('create table') and tag in ('CREATE FUNCTION')
   execute procedure test_event_trigger();
ERROR:  filter variable "tag" specified more than once
-- OK
create event trigger regress_event_trigger2 on ddl_command_start
   when tag in ('create table', 'CREATE FUNCTION')
   execute procedure test_event_trigger();
-- OK
comment on event trigger regress_event_trigger is 'test comment';
-- should fail, event triggers are not schema objects
comment on event trigger wrong.regress_event_trigger is 'test comment';
ERROR:  event trigger name cannot be qualified
-- drop as non-superuser should fail
create role regression_bob;
set role regression_bob;
create event trigger regress_event_trigger_noperms on ddl_command_start
   execute procedure test_event_trigger();
ERROR:  permission denied to create event trigger "regress_event_trigger_noperms"
HINT:  Must be superuser to create an event trigger.
reset role;
-- all OK
alter event trigger regress_event_trigger enable replica;
alter event trigger regress_event_trigger enable always;
alter event trigger regress_event_trigger enable;
alter event trigger regress_event_trigger disable;
-- regress_event_trigger2 and regress_event_trigger_end should fire, but not
-- regress_event_trigger
create table event_trigger_fire1 (a int);
NOTICE:  test_event_trigger: ddl_command_start CREATE TABLE
NOTICE:  test_event_trigger: ddl_command_end CREATE TABLE
-- regress_event_trigger_end should fire here
drop table event_trigger_fire1;
NOTICE:  test_event_trigger: ddl_command_end DROP TABLE
-- alter owner to non-superuser should fail
alter event trigger regress_event_trigger owner to regression_bob;
ERROR:  permission denied to change owner of event trigger "regress_event_trigger"
HINT:  The owner of an event trigger must be a superuser.
-- alter owner to superuser should work
alter role regression_bob superuser;
alter event trigger regress_event_trigger owner to regression_bob;
-- should fail, name collision
alter event trigger regress_event_trigger rename to regress_event_trigger2;
ERROR:  event trigger "regress_event_trigger2" already exists
-- OK
alter event trigger regress_event_trigger rename to regress_event_trigger3;
-- should fail, doesn't exist any more
drop event trigger regress_event_trigger;
ERROR:  event trigger "regress_event_trigger" does not exist
-- should fail, regression_bob owns regress_event_trigger2/3
drop role regression_bob;
ERROR:  role "regression_bob" cannot be dropped because some objects depend on it
DETAIL:  owner of event trigger regress_event_trigger3
-- these are all OK; the second one should emit a NOTICE
drop event trigger if exists regress_event_trigger2;
drop event trigger if exists regress_event_trigger2;
NOTICE:  event trigger "regress_event_trigger2" does not exist, skipping
drop event trigger regress_event_trigger3;
drop event trigger regress_event_trigger_end;
drop function test_event_trigger();
-- now test the sql_drop event trigger
create function test_event_trigger_for_sql_drop() returns event_trigger as $$
DECLARE
    obj record;
BEGIN
    RAISE NOTICE 'test_event_trigger: % %', tg_event, tg_tag;

    FOR obj IN SELECT * FROM pg_dropped_objects()
    LOOP
        -- we can't output the full data that we have here because the OID
        -- would change each time we run the regression tests.
        --
        -- obj.classId, obj.objid, obj.objsubid;
        RAISE NOTICE 'sql_drop event trigger: %', obj.classId::regclass;
    END LOOP;
END
$$ language plpgsql;
-- OK
create event trigger regress_event_trigger_sql_drop on sql_drop
   execute procedure test_event_trigger_for_sql_drop();
-- a simple enough test: cascade
create table evt_a(id serial);
create view evt_a_v as select id from evt_a;
drop table evt_a cascade;
NOTICE:  drop cascades to view evt_a_v
NOTICE:  test_event_trigger: sql_drop DROP TABLE
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_rewrite
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_class
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_class
NOTICE:  sql_drop event trigger: pg_class
-- another test with multiple targets
create table evt_a(id serial);
create table evt_b(id serial);
drop table evt_a, evt_b;
NOTICE:  test_event_trigger: sql_drop DROP TABLE
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_class
NOTICE:  sql_drop event trigger: pg_class
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_type
NOTICE:  sql_drop event trigger: pg_class
NOTICE:  sql_drop event trigger: pg_class
-- cleanup the sql_drop tests
drop event trigger if exists regress_event_trigger_sql_drop;
drop function test_event_trigger_for_sql_drop();
drop role regression_bob;
